mgr/manager.go

```diff
@@ -155,13 +158,112 @@ var validExtensions = map[string]bool{
 	".yml":  true,
 }
 
+// CheckDiskPKI checks the PKI information on disk against cert spec and alerts upon differences
+// Specifically, it checks that private key on disk matches spec algorithm & keysize,
+// and certificate  on disk matches CSR spec info
+func (m *Manager) CheckDiskPKI() error {
+
+	if m.Certs == nil || len(m.Certs) == 0 {
+		return nil
+	}
+
+	// Iterate through certificates. Compare what's on disk to what's structurally stored
+	for i := range m.Certs {
+		csm := m.Certs[i]
+		certPath := csm.Spec.Cert.Path
+		keyPath := csm.Spec.Key.Path
+		specPath := csm.Spec.Path
+		csrRequest := csm.Spec.Request
+
+		// Read private key algorithm and keysize from disk, determine if RSA or ECDSA
+		keyData, err := ioutil.ReadFile(keyPath)
+		if err != nil {
+			return err
+		}
+		pemKey, _ := pem.Decode(keyData)
+		if pemKey == nil {
+			return errors.New("Unable to pem decode private key on disk")
+		}
+
+		var algDisk string
+		var sizeDisk int
+		privKey, err := x509.ParsePKCS1PrivateKey(pemKey.Bytes)
+		if err != nil {
+			privKey, err := x509.ParseECPrivateKey(pemKey.Bytes)
+			if err != nil {
+				// If we get here, then invalid key type
+				return errors.New("manager: Unable to parse private key algorithm from disk")
+			}
+			// If we get here, then it's ECDSA
+			algDisk = "ecdsa"
+			sizeDisk = privKey.Curve.Params().BitSize
+		} else {
+			//If we get here, then it's RSA
+			algDisk = "rsa"
+			sizeDisk = privKey.N.BitLen()
+		}
+
+		// Check algorithm and keysize of private key on disk against what's defined in spec
+		algSpec := csrRequest.KeyRequest.Algo()
+		sizeSpec := csrRequest.KeyRequest.Size()
+
+		if algDisk != algSpec {
+			metrics.AlgorithmMismatchCount.WithLabelValues(specPath).Set(1)
+			fmt.Printf("ALERT: disk alg is %s but spec alg is %s\n", algDisk, algSpec)
+		} else {
+			metrics.AlgorithmMismatchCount.WithLabelValues(specPath).Set(0)
+		}
+
+		if sizeDisk != sizeSpec {
+			metrics.KeysizeMismatchCount.WithLabelValues(specPath).Set(1)
+			fmt.Printf("ALERT: disk key size is %d but spec key size is %d\n", sizeDisk, sizeSpec)
```

Comment:
Should this use the logger instead? Same for the other warnings below.

Reply:
How y'all usually check logs for certmgr? This print statement shows up in journalctl so I thought it was sufficient -- plus, the print statement is useful because it prints the error after running `certmgr ensure`

Reply:
logging is a requirement although I've not audited the code to enforce it- if we ever change the logging backend to something like logrus (which allows structured- json- logging), printf's like this will break the log stream.

Reply:
Alright, based on this conversation I'll change these print statements to use log.Errorf instead. Thanks!